pkg/reconciler/scheduling/placement/placement_reconcile_scheduling.go

```diff
@@ -18,456 +18,137 @@ package placement
-		//              requeue duration. But better than not requeueing at all.
-		r.enqueueAfter(clusterName, ns, time.Minute*2)
-		return reconcileStatusContinue, ns, nil
+	if len(placement.Spec.LocationWorkspace) > 0 {
+		locationWorkspace = logicalcluster.New(placement.Spec.LocationWorkspace)
+	} else {
+		locationWorkspace = logicalcluster.From(placement)
 	}
 
-	// sort found bindings, take the first one, just to have stability.
-	// TODO(sttts): somehow avoid the situation of multiple workload bindings. Admissions?
-	sort.Slice(workloadBindings, func(i, j int) bool {
-		return workloadBindings[i].Name < workloadBindings[j].Name
-	})
-	binding := workloadBindings[0]
-	negotiationClusterName := logicalcluster.New(binding.Spec.Reference.Workspace.Path)
-
-	locations, err := r.listLocations(negotiationClusterName)
+	validLocationNames, err := r.validLocationNames(placement, locationWorkspace)
 	if err != nil {
-		return reconcileStatusStop, nil, err
-	}
+		conditions.MarkFalse(placement, schedulingv1alpha1.PlacementReady, schedulingv1alpha1.LocationNotFoundReason, conditionsv1alpha1.ConditionSeverityError, err.Error())
+		return reconcileStatusContinue, placement, err
+	}
+
+	switch placement.Status.Phase {
+	case schedulingv1alpha1.PlacementBound:
+		// if selected location becomes invalid when placement is in bound state, set PlacementSynced
+		// to false.
+		if !isLocationSelected(placement, locationWorkspace, validLocationNames) {
+			conditions.MarkFalse(
+				placement,
+				schedulingv1alpha1.PlacementReady,
+				schedulingv1alpha1.LocationInvalidReason,
+				conditionsv1alpha1.ConditionSeverityError,
+				"Selected location is invalid for current placement",
+			)
+			return reconcileStatusContinue, placement, nil
+		}
+
+		conditions.MarkTrue(placement, schedulingv1alpha1.PlacementReady)
+		return reconcileStatusContinue, placement, nil
+	case schedulingv1alpha1.PlacementUnbound:
+		if isLocationSelected(placement, locationWorkspace, validLocationNames) {
+			// if the selected location is valid, keep it.
+			conditions.MarkTrue(placement, schedulingv1alpha1.PlacementReady)
+			return reconcileStatusContinue, placement, nil
+		}
+	}
+
+	// now it is pending state or in unbound state and needs a reselection
+	if validLocationNames.Len() == 0 {
+		placement.Status.Phase = schedulingv1alpha1.PlacementPending
+		placement.Status.SelectedLocation = nil
+		conditions.MarkFalse(
+			placement,
+			schedulingv1alpha1.PlacementReady,
+			schedulingv1alpha1.LocationNotFoundReason,
+			conditionsv1alpha1.ConditionSeverityError,
+			"No valid location is found")
+		return reconcileStatusContinue, placement, nil
+	}
+
+	candidates := make([]string, 0, validLocationNames.Len())
+	for loc := range validLocationNames {
+		candidates = append(candidates, loc)
+	}
+
+	chosenLocation := candidates[rand.Intn(len(candidates))]
+	placement.Status.SelectedLocation = &schedulingv1alpha1.LocationReference{
+		Path:         locationWorkspace.String(),
```

Comment:
or local? Should we set the Path to local path or leave it empty?

Reply:
nevermind. You set locationWorkspace in the local case.

Reply:
it is local if not set in spec. It was done at line45